### PR TITLE
feat(efs): import access point - `fromAccessPointAttributes()`

### LINKED ISSUE #10711: [aws-efs] Cannot import an existing EFS Access Point to use with a lambda function
Cannot use `efs.AccessPoint.fromAccessPointId()` with `lambda.FileSystem.fromEfsAccessPoint()`. the former returns an `IAccessPoint` when the later expect an `AccessPoint`. I think following the CDK guidelines, `lambda.FileSystem.fromEfsAccessPoint()` should expect an `IAccessPoint`, not an `AccessPoint`.

```
Argument of type 'IAccessPoint' is not assignable to parameter of type 'AccessPoint'.
```

### Reproduction Steps

```ts
filesystem: lambda.FileSystem.fromEfsAccessPoint(
  efs.AccessPoint.fromAccessPointId(this, 'ap', props.ap.accessPointId),
  '/mnt/efs',
),
```

### What did you expect to happen?

I would expect the CDK to let me use an existing VPC, FileSytem and AccessPoint for my new Lambda function.

I would also expect the `AccessPoint` class to provide a `fromAccessPointAttributes()` method to import the data that would accept `AccessPointId` and `FileSystem` attributes.

### What actually happened?

I was not able to import an existing `AccessPoint`. I ended up creating a new `AccessPoint` within the same stack as my lambda function.

### Environment

  - **CLI Version      :** 1.66.0
  - **Framework Version:** 1.66.0
  - **Node.js Version:** 12.16.3
  - **OS               :** Linux Ubuntu 18
  - **Language (Version):** TypeScript (3.8.3)

### Other

I would expect to have a `efs.AccessPoint.fromAccessPointAttributes()` to import an Access Point. The `fromAccessPointId()` doesn't actually import the `FileSystem` and that's an issue when using it in the lambda function.

Here is an example of how it would look like:

```ts
interface LambdaStackProps extends StackProps {
  vpcId: string;
  publicSubnetsId: string[];
  privateSubnetIds: string[];
  availabilityZones: string[];
  efs: efs.IFileSystem;
  ap: efs.IAccessPoint;
  sg: ISecurityGroup;
}

class LambdaStack extends Stack {
  fn: lambda.IFunction;
  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    let vpc = Vpc.fromVpcAttributes(this, 'vpc', {
      vpcId: props.vpcId,
      availabilityZones: props.availabilityZones,
      publicSubnetIds: props.publicSubnetsId,
      privateSubnetIds: props.privateSubnetIds,
    });

    let fileSystem = lambda.FileSystem.fromEfsAccessPoint(
      efs.AccessPoint.fromAccessPointAttributes(this, 'ap', {
        accessPointId: props.ap.accessPointId,
        fileSystem: efs.FileSystem.fromFileSystemAttributes(this, 'efs', {
          fileSystemId: props.efs.fileSystemId,
          securityGroup: SecurityGroup.fromSecurityGroupId(this, 'sg', props.sg.securityGroupId),
        }),
      }),
      '/mnt/efs',
    );

    this.fn = new lambda.Function(this, 'lambda', {
      runtime: lambda.Runtime.NODEJS_12_X,
      code: lambda.Code.fromInline('exports.handler = function(event, ctx, cb) { return cb(null, "hi"); }'),
      handler: 'index.handler',
      vpc: vpc,
      filesystem: fileSystem,
    });
  }
}
```
--- 

This is :bug: Bug Report

* [x] :wave: I may be able to implement the fix
* [x] :warning: This fix might incur a breaking change

### PR DESCRIPTION
**Pull request by DaWyz**
Cannot use `efs.AccessPoint.fromAccessPointId()` with `lambda.FileSystem.fromEfsAccessPoint()`. the former returns an `IAccessPoint` when the later expect an `AccessPoint`. I think following the CDK guidelines, `lambda.FileSystem.fromEfsAccessPoint()` should expect an `IAccessPoint`, not an `AccessPoint`.

Argument of type `IAccessPoint` is not assignable to parameter of type `AccessPoint`.

### Solution
----

Add a new import method to the `AccessPoint` class called `fromAccessPointAttributes()` allowing to pass a fileSystem as an attribute.

Closes #10711.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by DaWyz** [comment]
I just looked into it again. I'm not sure it makes sense to me anymore. Also, the issue is still valid, the implementation I made doesn't seem like the right way to do it. If you think the same, I'm happy to discuss about the implementation and close this Pull Request.

I'm just confused about the `FileSystem` class in the `@aws-cdk/aws-lambda` and I'm wondering if it's not kind of redundant.

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (shivlaks) [inline packages/@aws-cdk/aws-efs/lib/access-point.ts:173]: we should extract this to a class and return that type - here's an example in rds [`fromDatabaseClusterAttributes`](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-rds/lib/cluster.ts#L463-L465) for [`ImportedDatabaseCluster`](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-rds/lib/cluster.ts#L368) for inspiration

**Review by shivlaks — DISMISSED** [review]
@DaWyz 

awesome first pass!! thanks for getting the ball rolling on this. I have some thoughts that I'd love to work in before merging this in. The most important one is creating a class that represents an imported file system. Take a look at the example of how it's done in `rds` and let me know what you think!

**Comment by DaWyz** [comment]
> @DaWyz
> 
> awesome first pass!! thanks for getting the ball rolling on this. I have some thoughts that I'd love to work in before merging this in. The most important one is creating a class that represents an imported file system. Take a look at the example of how it's done in `rds` and let me know what you think!

@shivlaks We could extract the [existing imported file system class](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-efs/lib/efs-file-system.ts#L214-L224) if it's what you mean. I can do the same for the AccessPoint as well.

About the IFileSystem property in the IAccessPoint, I made it mandatory because of the [lambda efs implementation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/filesystem.ts#L53-L79). This is the only place where the AccessPoint class is used and it needs the FileSystem to be able to setup the ingress rules.

So, at the moment, we cannot import an existing access point since it doesn't have the fileSystem property when using `fromAccessPointId`. I guess we could make the `FileSystem` property of `fromAccessPointAttributes` optional by updating the code in the [lambda efs implementation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/filesystem.ts#L53-L79) to check for `ap.fileSystem` presence and make if fail with a nice error message in case it's missing.

Also, I dropped the existing `fromAccessPointId` method and I don't think I should. Making the `fileSystem` property optional would allow me to add `fromAccessPointId` method back.

Let me know what you think and I can update the code accordingly.

**Comment by shivlaks** [comment]
> @shivlaks We could extract the [existing imported file system class](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-efs/lib/efs-file-system.ts#L214-L224) if it's what you mean. I can do the same for the AccessPoint as well.

yep that is what I mean. extract to `ImportedFileSystem` or so that implements IFileSystem and we just return new `ImportedFileSystem` in the `fromFileSystemAttributes` method
 
> About the IFileSystem property in the IAccessPoint, I made it mandatory because of the [lambda efs implementation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/filesystem.ts#L53-L79). This is the only place where the AccessPoint class is used and it needs the FileSystem to be able to setup the ingress rules.

got it, thanks for explaining! might be worth a comment.

> So, at the moment, we cannot import an existing access point since it doesn't have the fileSystem property when using `fromAccessPointId`. I guess we could make the `FileSystem` property of `fromAccessPointAttributes` optional by updating the code in the [lambda efs implementation](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/filesystem.ts#L53-L79) to check for `ap.fileSystem` presence and make if fail with a nice error message in case it's missing.

It seems reasonable to me. Adding the `fileSystem` property would have to be optional as you've suggested because the alternative is a breaking change and the `lambda` module is stable.
I'd defer to @nija-at on whether that's an approach he's comfortable with as he's the primary maintainer of the module.


> Also, I dropped the existing `fromAccessPointId` method and I don't think I should. Making the `fileSystem` property optional would allow me to add `fromAccessPointId` method back.

I agree, let's avoid dropping the method.

> Let me know what you think and I can update the code accordingly.

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/lib/access-point.ts:119]: This API should not be removed. Keep the `fromAccessPointId()` alongside `fromAccessPointAttributes()`.

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/lib/access-point.ts:27]: How about make this mandatory and throw an error in the implementation of `fromAccessPointId()`? 

The error message can be "fileSystemId is not available when 'fromAccessPointId()' is used. Use 'fromAccessPointAttributes()' instead"

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/lib/access-point.ts:125]: Not true. It should be

```suggestion
   * @default - determined based on accessPointArn
```

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/lib/access-point.ts:133]: Same comment as above.

File: packages/@aws-cdk/aws-efs/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/README.md:74]: Make this part clear. What does it mean to "make sure the mount targets are available"?

File: packages/@aws-cdk/aws-efs/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/README.md:78]: Import implies that you're bringing in an access point that was created outside the CDK app. You've already stated this as the first sentence.

Drop this last sentence entirely.

File: packages/@aws-cdk/aws-efs/README.md
Comment (nija-at) [inline packages/@aws-cdk/aws-efs/README.md:77]: Some small formatting adjustments.

```suggestion
⚠️ Notice: When importing an Access Point using `fromAccessPointAttributes()`, you must make sure the mount targets are
available. Otherwise, you may encounter the following error when deploying:

> EFS file system <ARN of efs> referenced by access point <ARN of access point of EFS> has
mount targets created in all availability zones the function will execute in, but not all are in the available life cycle
state yet. Please wait for them to become available and try the request again.
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Some suggestions below.

File: packages/@aws-cdk/aws-efs/lib/access-point.ts
Comment (DaWyz) [reply packages/@aws-cdk/aws-efs/lib/access-point.ts:27]: @nija-at, that would also be a breaking change as well, right ?

I'm wondering if there is a use-case (outside of EFS for lambda) where we would need to import an AccessPoint without having the need for `FileSystem` ?

Comment (nija-at) [reply packages/@aws-cdk/aws-efs/lib/access-point.ts:27]: efs is an experimental module. Breaking changes are acceptable as long as they are [called out in the commit description](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#step-4-commit).

> I'm wondering if there is a use-case (outside of EFS for lambda) where we would need to import an AccessPoint without having the need for FileSystem ?

In such cases exist, the `fromAccessPointId()` API can be used.

Comment (DaWyz) [reply packages/@aws-cdk/aws-efs/lib/access-point.ts:27]: I see. I didn't get the idea of how to implement this at start. I think I'm good now. I'm doing the same as in the rds imported cluster and throw the exception on the Getter when fileSystem is not present.. See below. Will update the PR.

```ts
class ImportedAccessPoint extends AccessPointBase {
  public readonly accessPointId: string;
  public readonly accessPointArn: string;
  private readonly _fileSystem?: IFileSystem;

  constructor(scope: Construct, id: string, attrs: AccessPointAttributes) {
    super(scope, id);

    if (!attrs.accessPointId) {
      if (!attrs.accessPointArn) {
        throw new Error('One of accessPointId or AccessPointArn is required!');
      }

      this.accessPointArn = attrs.accessPointArn;
      let maybeApId = Stack.of(scope).parseArn(attrs.accessPointArn).resourceName;

      if (!maybeApId) {
        throw new Error('ARN for AccessPoint must provide the resource name.');
      }

      this.accessPointId = maybeApId;
    } else {
      if (attrs.accessPointArn) {
        throw new Error('Only one of accessPointId or AccessPointArn can be provided!');
      }

      this.accessPointId = attrs.accessPointId;
      this.accessPointArn = Stack.of(scope).formatArn({
        service: 'elasticfilesystem',
        resource: 'access-point',
        resourceName: attrs.accessPointId,
      });
    }

    this._fileSystem = attrs.fileSystem;
  }

  public get fileSystem() {
    if (!this._fileSystem) {
      throw new Error("fileSystem is not available when 'fromAccessPointId()' is used. Use 'fromAccessPointAttributes()' instead");
    }

    return this._fileSystem;
  }
}
```

**Review by shivlaks — APPROVED** [review]
@DaWyz - changes look good to me!
i added a `do-not-merge` so @nija-at can also take a look

**Comment by shivlaks** [comment]
confirmed with @nija-at that we shouldn't wait on another review, thanks for the contribution @DaWyz !!

**Comment by DaWyz** [comment]
> confirmed with @nija-at that we shouldn't wait on another review, thanks for the contribution @DaWyz !!

No problem! Happy to help! I'm sure you guys are extremely busy with CDK so thank you for taking the time to review the PRs! Keep up the good work!

**Pull request merged**
